### PR TITLE
update validation functions

### PR DESCRIPTION
**Pull request by IanAtCredo**
## Describe your changes
`TabularData` wasn't properly validating pre/post processing. This PR fixes that. 

Also added basic name validation.

## Known outstanding issues that are not fully accounted for
* Should we just require the input to be a dataframe?

## Checklist before requesting a review
- [x] I have performed a self-review of my code
- [ ] I have built basic tests for new functionality (particularly new evaluators)
- [ ] If new libraries have been added, I have checked that [readthedocs](https://readthedocs.org/projects/credoai-lens/) API documentation is constructed correctly
- [ ] Will this be part of a major product update? If yes, please write one phrase about this update.

## Extra-mile Checklist
- [ ] I have thought expansively about edge cases and written tests for them

### CODE REVIEW

File: credoai/artifacts/data/tabular_data.py [hunk 1]
```diff
@@ -52,16 +52,20 @@ def copy(self):
...
 
     def _process_y(self, y):
-        """Standardize y data"""
-        # if X is pandas object, and y is convertible, convert y to
-        # pandas object with X's index
+        """Standardize y data
+
+        If y is convertible, convert y to pandas object with X's index
+        """
         if isinstance(y, (np.ndarray, list)):
```
Comment (esherman-credo) [inline credoai/artifacts/data/tabular_data.py:69]: What happens if this `isinstance()` call returns False? That would be the case if the user passed something entirely ill-formed --> seems like right now we're just assuming Y matches the docstring and is thus always a DF/Series, nd.Array, or List?

File: credoai/artifacts/data/tabular_data.py [hunk 2]
```diff
@@ -72,25 +76,30 @@ def _process_y(self, y):
                 y = pd_type(y, name="target")
         return y
 
+    def _validate_X(self):
+        try:
+            pd.DataFrame(self.X)
```
Comment (esherman-credo) [inline credoai/artifacts/data/tabular_data.py:81]: Technically someone could pass a dictionary and this wouldn't fail. They'd be violating the documentation but I _think_ it'd still end up in the right format based on other validation. Nevertheless, it might be better to explicitly forbid it through `isinstance` checking since there could be unanticipated issues with dictionaries.

Example:
```
>>> blarg = {'col1': [4, 5, 6], 
... 'col2': [1, 2, 3]}
>>> import pandas as pd
>>> pd.DataFrame(blarg)
   col1  col2
0     4     1
1     5     2
2     6     3
```

File: credoai/artifacts/data/tabular_data.py [hunk 3]
```diff
@@ -72,25 +76,30 @@ def _process_y(self, y):
...
+            raise ValidationError(
+                "X and y are not the same length. "
+                + f"X Length: {len(self.X)}, y Length: {len(self.y)}"
+            )
+
+    def _validate_processed_X(self):
+        """Validate processed X"""
+        if len(self.X.columns) != len(set(self.X.columns)):
+            raise ValidationError("X contains duplicate column names")
+        if len(self.X.index) != len(set(self.X.index)):
```
Comment (esherman-credo) [inline credoai/artifacts/data/tabular_data.py:97]: It would make sense to also check that `sensitive_feature.columns` doesn't intersect with `X.columns` if both objects are specified.

**Review by esherman-credo — CHANGES_REQUESTED** [review]
Some extra validation is needed for y.

Other comments are nice-to-haves.

File: credoai/artifacts/data/tabular_data.py (hunk 3 above)
Comment (IanAtCredo) [reply credoai/artifacts/data/tabular_data.py:97]: I don't think that's something we want to forbid. The sensitive features are used for sensitive feature things - they could exist in X as well (to be passed to the model)

File: credoai/artifacts/data/tabular_data.py (hunk 2 above)
Comment (IanAtCredo) [reply credoai/artifacts/data/tabular_data.py:81]: I looked for validation of "array-like" as requested by sklearn. It doesn't seem like there is such validation. This seems like quite an odd case to be passed and not revealed by other validations

**Comment by IanAtCredo** [comment]
@esherman-credo , found a sklearn util called `check_array` which seems almost like what we want. Added onto it to address your comments.

**Review by esherman-credo — COMMENTED** [review]
Changes look good. I think we need to think harder about what to do about overlap between sensitive features and X.

File: credoai/artifacts/data/tabular_data.py (hunk 3 above)
Comment (IanAtCredo) [reply credoai/artifacts/data/tabular_data.py:94]: It won't do deceptively well... it will do really well and that is what we want. The features used for `sensitive_features` are distinct from the features used in X, which is the predictor features.

**Review by esherman-credo — APPROVED** [review]
